### ISSUE TITLE
Add viewer count preference when using CDNs

### ISSUE THREAD

**Issue by gabek** [owncast/owncast#2627]
### Share your bug report, feature request, or comment.

- Current approach: Count both people watching via the web interface and those using external players. This uses the IP address of the viewer as an identifier and is less accurate when a CDN is between the viewer and the Owncast server, as Owncast might see a single viewer coming from multiple IP addresses and count them as multiple viewers.
- Count only the number of people watching via the web interface. This number is more accurate but excludes all external viewers. This would useful when enabling a CDN.

As an addition to option 2 I've thought about possibly setting a cookie the first time you request video assets or make a `ping` request and use that as an identifier, but 3rd party players can make no guarantees that they can use cookies. Cookies outside of web browsers are a mess. Even cookies inside browsers are a mess and I'd prefer not to have any. But I'll do some experimenting in the future to find a best-effort solution

**Comment by jprjr** [comment]
> * Current approach: Count both people watching via the web interface and those using external players. This uses the IP address of the viewer as an identifier and is less accurate when a CDN is between the viewer and the Owncast server, as Owncast might see a single viewer coming from multiple IP addresses and count them as multiple viewers.

One thing I noticed when using s3 storage is, the main playlist is served by Owncast, and variant stream playlists are served by s3, meaning for external players, Owncast is only notified about that first GET request.

I think the viewers are calculated by pulling playlists - so what about having owncast serve all playlists, and only uploading video segments to s3? I think this would require putting the full segment URL in the playlists, not just the segment filename. And for CDN, you could specify that only video segments should be CDN'd.

**Comment by gabek** [comment]
> I think the viewers are calculated by pulling playlists

Right now it counts by two different methods and combines them. (Using IP address to count as a unique client)

1. HLS video requests.
2. The `/api/ping` endpoint. It simply tells Owncast that this viewer is still watching.

Between the two it covers most cases. If you use object storage, but using the Owncast provided player, then you're counted. If you're not using the Owncast provided player, but not using object storage, you're counted. Where it falls short is if you're not using the Owncast provided player _and_ you're using object storage.

I wouldn't feel comfortable pulling the playlists directly from Owncast because of how often they're queried. If the goal of using object storage is to take the load off the Owncast server, then making n number of clients hit Owncast every 3 seconds or whatever, it kind of defeats the point. It's not just about bandwidth, but number of concurrent requests. Even for a tiny file.

**Comment by jprjr** [comment]
Sorry, just to clarify - I meant to say, I think (and I could be wrong) that viewers are calculated specifically from requests for the HLS playlists (and not the video segments). I think the phrase "HLS video requests" is a bit vague - I could see that as meaning either playlists or video segments (and maybe it does mean either?). As well as the `/api/ping` endpoint you mentioned.

I'd argue having the playlists come from owncast and media segments from external storage wouldn't really defeat the point, it highly depends on the streamer's circumstances. The reason I switched to object storage was for bandwidth reasons. But, splitting it out would be a more complex setup, and odds are there's not that many external players viewing the stream anyway.

**Comment by gabek** [comment]
Yes, specifically it's the playlists that register the users. And while I think bandwidth is likely the most common reason to use object storage, I do think quantity of requests overall is a reasonable reason to switch to object storage as well. With enough viewers HLS could turn into a DDoS. If we changed this then people who have a very large audience wouldn't be able to take advantage of Owncast due to the sheer number of requests, right now object storage solves that problem without needing a CDN (and adding a CDN adds its own viewer count issues as originally stated).

> and odds are there's not that many external players viewing the stream anyway.

That's what I think too, so that's why I've been ok with these viewer count trade offs so far.
